### PR TITLE
fix(ses): Add Node.js domain hazard mitigation

### LINKED ISSUE #126: Node.js adds a property called `domain` to promises
Mark and I noticed that Node.js adds a property called `domain` to promises while running something from a REPL in SwingSet. 

Mark says that `domain` leads to all sorts of objects whose semantics we haven't inspected for safety. This is an authority leak. If SES can't remove the domain property from all promises before the promise is exposed to user code, then this is a severe authority leak for SES as run on Node.js. Fortunately it seems to only be a problem when code is run from a REPL. 

[edit 2020-09-15 @kriskowal]

SES will not be secure on Node.js until we have the means to disable `domains`. We must work with the Node.js project to reveal a command-line flag or the moral equivalent.

### PR DESCRIPTION
**Pull request by kriskowal**
Fixes #126

### CODE REVIEW

**Comment by kriskowal** [comment]
ht @dominictarr for investigating the technique applied here.

**Comment by kriskowal** [comment]
Looks like this is going to have to sit around until we can remove support for [RESM](https://github.com/standard-things/esm), and this will be a breaking change.

**Review by erights — APPROVED** [review]
LGTM -- I hope we can merge this for real soon!

**Comment by erights** [comment]
At https://github.com/endojs/endo/pull/850/checks?check_run_id=3138995713
why would this fail by failing the sloppy check?

**Comment by kriskowal** [comment]
> At https://github.com/endojs/endo/pull/850/checks?check_run_id=3138995713
> why would this fail by failing the sloppy check?

This is an easy fix for a problem I’ve seen before. Instead of using `process` as a free variable, I need to use `globalThis.process`. That keeps Parcel from noticing the dependency on `process` and stops it from including a Node.js shim for that global which has the unintended side-effect of turning the bundle into a sloppy mode bundle.

The error with the -r esm packaging test is not solvable. We can’t land this until all our dependees can migrate off RESM.

File: packages/ses/error-codes/SES_NO_DOMAINS.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# SES failed to lockdown, Node.js domains have been initialized (SES_NO_DOMAINS)
```
Comment (erights) [inline packages/ses/error-codes/SES_NO_DOMAINS.md:1]: This text should also mention `domainTaming` and link to an explanation.

File: packages/ses/src/lockdown-shim.js [hunk 2]
```diff
@@ -139,6 +140,7 @@ export const repairIntrinsics = (
     overrideTaming = 'moderate',
     overrideDebug = [],
     stackFiltering = 'concise',
+    domainTaming = 'unsafe', // To become 'safe' by default in next-breaking-release.
```
Comment (erights) [inline packages/ses/src/lockdown-shim.js:143]: Put a `TODO` in this comment text?

File: packages/ses/src/tame-domains.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  // Protect against the hazard presented by Node.js domains.
+  if (typeof globalThis.process === 'object' && globalThis.process !== null) {
+    // Check whether domains were initialized.
+    const domainDescriptor = getOwnPropertyDescriptor(
+      globalThis.process,
+      'domain',
+    );
+    if (domainDescriptor !== undefined && domainDescriptor.get !== undefined) {
```
Comment (erights) [inline packages/ses/src/tame-domains.js:26]: Why the `get` test?

**Review by erights — APPROVED** [review]
Please add a section to lockdown-options.md describing this option. While you're there, please check that it does still document all options.

LGTM

File: packages/ses/src/tame-domains.js (hunk 3 above)
Comment (kriskowal) [reply packages/ses/src/tame-domains.js:26]: This is an important distinction and I’ve added a comment to the effect. Node.js initializes process with a process.domain === null and replaces this with a get/set pair when `domain` initializes.

**Pull request merged**